chirp/drivers/baofeng_1900.py

```diff
@@ -0,0 +1,596 @@
+
+    if ack != CMD_ACK:
+        raise Exception("No ACK reading block %04x." % (block_addr))
+
+    return block_data
+
+
+def _bf1900_write_block(radio, block_addr, block_size):
+    serial = radio.pipe
+
+    cmd = struct.pack(">cHb", b'W', block_addr, BLOCK_SIZE)
+    data = radio.get_mmap().get_byte_compatible()[block_addr:block_addr + 8]
+
+    LOG.debug("Writing Data:")
+    LOG.debug(util.hexprint(cmd + data))
+
+    try:
+        serial.write(cmd + data)
+        # Time required to write data blocks varies between individual
+        # radios of the Baofeng BF-888S model. The longest seen is
+        # ~0.31s.
+        if serial.read(1) != CMD_ACK:
+            raise Exception("No ACK")
+    except:
+        raise errors.RadioError("Failed to send block "
+                                "to radio at %04x" % block_addr)
+
+
+def do_download(radio):
+    LOG.debug("download")
+    _bf1900_enter_programming_mode(radio)
+
+    data = b""
+
+    status = chirp_common.Status()
+    status.msg = "Cloning from radio"
+
+    status.cur = 0
+    status.max = radio._memsize
+
+    for addr in range(0, radio._memsize, BLOCK_SIZE):
+        status.cur = addr + BLOCK_SIZE
+        radio.status_fn(status)
+
+        block = _bf1900_read_block(radio, addr, BLOCK_SIZE)
+        data += block
+
+        LOG.debug("Address: %04x" % addr)
+        LOG.debug(util.hexprint(block))
+
+    _bf1900_exit_programming_mode(radio)
+
+    return memmap.MemoryMapBytes(data)
+
+
+def do_upload(radio):
+    status = chirp_common.Status()
+    status.msg = "Uploading to radio"
+
+    _bf1900_enter_programming_mode(radio)
+
+    status.cur = 0
+    status.max = radio._memsize
+
+    for start_addr, end_addr in radio._ranges:
+        for addr in range(start_addr, end_addr, BLOCK_SIZE):
+            status.cur = addr + BLOCK_SIZE
+            radio.status_fn(status)
+            _bf1900_write_block(radio, addr, BLOCK_SIZE)
+
+    _bf1900_exit_programming_mode(radio)
+
+
+class BaofengBF1904(chirp_common.Alias):
+    VENDOR = 'Baofeng'
+    MODEL = 'BF-1904'
+
+
+@directory.register
+class BF1900Radio(chirp_common.CloneModeRadio):
+    VENDOR = "Baofeng"
+    MODEL = "BF-1901"
+    BAUD_RATE = 9600
+    NEEDS_COMPAT_SERIAL = False
+
+    ALIASES = [BaofengBF1904]
+    SIDEKEYFUNCTION_LIST = ["Off", "Monitor", "Transmit Power", "Alarm"]
+
+    # This code currently requires that ranges start at 0x0000
+    # and are continuous. In the original program 0x0388 and 0x03C8
+    # are only written (all bytes 0xFF), not read.
+    # _ranges = [
+    #       (0x0000, 0x0110),
+    #       (0x02B0, 0x02C0),
+    #       (0x0380, 0x03E0)
+    #       ]
+    # Memory starts looping at 0x1000... But not every 0x1000.
+
+    _ranges = [
+        (0x0000, 0x0110),
+        (0x0380, 0x03E0),
+        (0x02B0, 0x02C0),
+    ]
+    _memsize = 0x03E0
+    _has_fm = True
+    _has_sidekey = False
+    _has_scanmodes = True
+    _has_scramble = False
```

Comment:
I think you get the idea, but use flags like this to control the different behavior in things like `get_settings()` for your subclass so that the right things will get exposed (or not).